polygeist/tools/cgeist/Lib/clang-mlir.cc

```diff
@@ -1552,8 +1552,15 @@ ValueCategory MLIRScanner::CommonFieldLookup(clang::QualType CT,
 
     Result = Builder.create<polygeist::SubIndexOp>(Loc, ResultType, val,
                                                    getConstantIndex(fnum));
-  } else if (auto RT = MT.getElementType().dyn_cast<mlir::sycl::AtomicType>()) {
-    llvm_unreachable("not implemented");
+  } else if (auto AT = MT.getElementType().dyn_cast<mlir::sycl::AtomicType>()) {
+    assert(fnum < AT.getBody().size() && "ERROR");
+
+    const auto ElementType = AT.getBody()[fnum];
```

Comment:
According to the body, the element tyoe should be a memref, but according to the sycl to llvm conversion pass, the element type is actually a llvm pointer.

Reply:
I was thinking of not using the `body` (use just the Dataype and address space member vars) to create the ElementType here.  Do you think that will work for now?  Setting the body as llvm.ptr during generation of atomic op does not generalize well.  However, if we are using the Datatype and addressspace to create the `ElementType` here, having the body is redundant.  Any thoughts?

Reply:
I gave a bit more thoughts. We can continue to use memref, and improve convertToLLVMABI pass to change the memref to llvm pointer.

Reply:
Thanks.  Does it mean that in the `SYCLToLLVM` pass, we can use `convertBodyType` to generate the 5-member struct that is result of conversion from the `memref<?xi32, 1>` and clean up the struct to match clang generated code in the `convertToLLVMABI` pass?

Also just curious, do you plan to convert all `memref`s to `llvm.ptr` in `convertToLLVMABI` pass or just a subset of the `memref`s?

Reply:
convertToLLVMABI should happens before SYCLToLLVM, so SYCLToLLVM shouldn't receive the body of an atomic type as memref. 
Currently, convertToLLVMABI pass converts all memref arguments to llvm pointer. We can improve it to convert sycl type's body memref to llvm pointer as well.